hbase-common/src/main/java/org/apache/hadoop/hbase/nio/MultiByteBuff.java

```diff
@@ -1064,28 +1065,58 @@ public void get(ByteBuffer out, int sourceOffset, int length) {
     return output;
   }
 
-  @Override
-  public int read(ReadableByteChannel channel) throws IOException {
+  private int internalRead(ReadableByteChannel channel, long offset,
+      ChannelReader reader) throws IOException {
     checkRefCount();
     int total = 0;
     while (true) {
-      // Read max possible into the current BB
-      int len = channelRead(channel, this.curItem);
-      if (len > 0)
+      int len = read(channel, this.curItem, offset, reader);
```

Comment:
Seems we can also create a iterator to read & fill the ByteBuff ?  Similar with the BufferIterator  in ByteBufferArray, I think that would be more clearer.... Can be a following issue ?